File: phonetics.py
"""
Gujlish phonetic normalisation.

The whole engine rests on one idea: romanised Gujarati has no standard
spelling. "chhe", "che" and "chey" are the same word. So we map every
surface form to a coarse phonetic key, index on the key, and display
the canonical surface form.

Two keys per word, because not all variation is equal:

  strict_key  keeps aspiration (th / dh / kh / gh / bh / jh / ph)
  loose_key   throws it away

People are nearly unanimous about writing "thayu" with the h, so
conflating th and t at the top level makes "tha" suggest "tame" ahead
of "thayu" -- wrong, and annoying. But some people do write "tayu", so
the loose key stays available as a scored-down fallback.

Prefix mode matters too. A whole word gets its trailing nasal stripped
("chhun" -> "cu") because nasalisation is written inconsistently. A
half-typed prefix must not: stripping the m from "jam" gives "ja" and
buries "jaman" under every word starting with ja-.

This is the reference implementation, ported to Swift for the keyboard.
Dependency-free and deliberately dull: no regex, no unicode tricks.
"""

# Aspirated consonants get a private uppercase symbol so later rules
# can't touch them and so the loose key can undo them in one pass.
_DIGRAPHS = [
    ("chh", "C"),
    ("ch", "C"),
    ("shh", "s"),
    ("sh", "s"),
    ("th", "T"),
    ("dh", "D"),
    ("kh", "K"),
    ("gh", "G"),
    ("ph", "F"),
    ("bh", "B"),
    ("jh", "J"),
    ("zh", "J"),
]

_SINGLES = {"z": "J", "w": "v", "q": "k", "f": "F"}

_VOWEL_RUNS = [
    ("aa", "a"), ("ee", "i"), ("ii", "i"),
    ("oo", "u"), ("uu", "u"), ("ei", "e"),
]

# strict symbol -> loose equivalent
_DEASPIRATE = {
    "T": "t", "D": "d", "K": "k", "G": "g",
    "F": "f", "B": "b", "J": "j", "C": "c",
}

VOWELS = set("aeiou")


def _core(word, prefix):
    w = "".join(c for c in word.lower() if c.isalpha())
    if not w:
        return ""

    w = w.replace("x", "ks")

    for src, dst in _DIGRAPHS:
        w = w.replace(src, dst)

    w = "".join(_SINGLES.get(c, c) for c in w)

    # y is a vowel everywhere but word-initially: thayu/thaiu, chey/che,
    # kay/kai all need to land together.
    if len(w) > 1:
        w = w[0] + w[1:].replace("y", "i")

    for _ in range(3):
        before = w
        for src, dst in _VOWEL_RUNS:
            w = w.replace(src, dst)
        if w == before:
            break

    # Doubled consonants: sachchu / sacchu / sachu. Compare
    # case-insensitively so c+C collapses to the aspirated form.
    out = []
    for c in w:
        if out and out[-1].lower() == c.lower() and c.lower() not in VOWELS:
            if c.isupper():
                out[-1] = c
            continue
        out.append(c)
    w = "".join(out)

    if not prefix:
        # Whole words only. Nasalisation and a trailing bare h carry no
        # information: chhu/chhun/chun, ha/haan/han.
        while len(w) > 2 and w[-1] in "nm" and w[-2] in VOWELS:
            w = w[:-1]
        while len(w) > 2 and w[-1] == "h":
            w = w[:-1]

    return w


def strict_key(word, prefix=False):
    """Phonetic key that preserves aspiration."""
    return _core(word, prefix)


def loose_key(word, prefix=False):
    """Phonetic key that also folds th->t, kh->k, chh->ch and so on."""
    w = _core(word, prefix)
    return "".join(_DEASPIRATE.get(c, c) for c in w)


if __name__ == "__main__":
    groups = [
        ["che", "chhe", "chey", "chhey"],
        ["chu", "chhu", "chhun", "chun"],
        ["thayu", "thayoo", "thaiu"],
        ["su", "shu", "shoo", "soo"],
        ["majama", "majaama"],
        ["nathi", "nathee"],
        ["jamva", "jamwa", "jamvaa"],
        ["khabar", "khabbar"],
        ["saru", "saaru", "sarun"],
        ["kai", "kay", "kaai"],
        ["ghare", "gharey"],
        ["sacchu", "sachu", "sachchu"],
        ["aavjo", "avjo"],
        ["tame", "tamey", "tamme"],
        ["dikro", "deekro"],
        ["paisa", "paysa", "paisaa"],
        ["fafda", "phaphda", "faphda"],
        ["jhaju", "zaju"],
    ]
    bad = 0
    for g in groups:
        keys = {strict_key(w) for w in g}
        if len(keys) != 1:
            bad += 1
        print(f"{'ok ' if len(keys) == 1 else 'BAD'} {g[0]:<10} -> {sorted(keys)}")
    print(f"\n{len(groups) - bad}/{len(groups)} variant groups collapse on the strict key")

    print("\nAspiration kept apart on strict, folded on loose:")
    for a, b in [("thayu", "tayu"), ("khabar", "kabar"), ("ghare", "gare")]:
        print(f"    {a}/{b}: strict {strict_key(a)}/{strict_key(b)}"
              f"   loose {loose_key(a)}/{loose_key(b)}")

    print("\nPrefix mode must not strip the trailing nasal:")
    for p in ["jam", "kem", "chhu", "tam"]:
        print(f"    {p!r}: word {strict_key(p)!r}  prefix {strict_key(p, True)!r}")

File: ios/GujlishCore/Sources/GujlishCore/Phonetics.swift
// Gujlish phonetic normalisation. Port of phonetics.py, which is the
// reference: every surface form maps to a coarse phonetic key, the
// lexicon is indexed on the key, and the canonical surface is shown.
//
//   strictKey  keeps aspiration (th / dh / kh / gh / bh / jh / ph)
//   looseKey   throws it away
//
// Prefix mode keeps the trailing nasal and h, because a half-typed
// "jam" must not collapse to "ja". Deliberately dull: no regex.

public enum Phonetics {
    // Aspirated consonants get a private uppercase symbol so later rules
    // can't touch them and so the loose key can undo them in one pass.
    // Order matters: "chh" before "ch", "shh" before "sh".
    private static let digraphs: [([Character], [Character])] = [
        ("chh", "C"), ("ch", "C"), ("shh", "s"), ("sh", "s"),
        ("th", "T"), ("dh", "D"), ("kh", "K"), ("gh", "G"),
        ("ph", "F"), ("bh", "B"), ("jh", "J"), ("zh", "J"),
    ].map { (Array($0.0), Array($0.1)) }

    private static let singles: [Character: Character] = ["z": "J", "w": "v", "q": "k", "f": "F"]

    private static let vowelRuns: [([Character], [Character])] = [
        ("aa", "a"), ("ee", "i"), ("ii", "i"),
        ("oo", "u"), ("uu", "u"), ("ei", "e"),
    ].map { (Array($0.0), Array($0.1)) }

    // strict symbol -> loose equivalent
    private static let deaspirate: [Character: Character] = [
        "T": "t", "D": "d", "K": "k", "G": "g",
        "F": "f", "B": "b", "J": "j", "C": "c",
    ]

    private static let vowels: Set<Character> = ["a", "e", "i", "o", "u"]

    /// Phonetic key that preserves aspiration.
    public static func strictKey(_ word: String, prefix: Bool = false) -> String {
        String(core(word, prefix: prefix))
    }

    /// Phonetic key that also folds th->t, kh->k, chh->ch and so on.
    public static func looseKey(_ word: String, prefix: Bool = false) -> String {
        String(core(word, prefix: prefix).map { deaspirate[$0] ?? $0 })
    }

    // Left-to-right, non-overlapping, like Python's str.replace.
    private static func replace(_ w: [Character], _ src: [Character], _ dst: [Character]) -> [Character] {
        guard w.count >= src.count else { return w }
        var out: [Character] = []
        out.reserveCapacity(w.count)
        var i = 0
        let last = w.count - src.count
        while i < w.count {
            if i <= last && w[i] == src[0] && Array(w[i..<(i + src.count)]) == src {
                out.append(contentsOf: dst)
                i += src.count
            } else {
                out.append(w[i])
                i += 1
            }
        }
        return out
    }

    private static func core(_ word: String, prefix: Bool) -> [Character] {
        var w: [Character] = Array(word.lowercased()).filter { $0.isLetter }
        if w.isEmpty { return w }

        w = replace(w, ["x"], ["k", "s"])

        for (src, dst) in digraphs {
            w = replace(w, src, dst)
        }

        w = w.map { singles[$0] ?? $0 }

        // y is a vowel everywhere but word-initially: thayu/thaiu, chey/che,
        // kay/kai all need to land together.
        for i in w.indices.dropFirst() where w[i] == "y" {
            w[i] = "i"
        }

        for _ in 0..<3 {
            let before = w
            for (src, dst) in vowelRuns {
                w = replace(w, src, dst)
            }
            if w == before { break }
        }

        // Doubled consonants: sachchu / sacchu / sachu. Compare
        // case-insensitively so c+C collapses to the aspirated form.
        var out: [Character] = []
        out.reserveCapacity(w.count)
        for c in w {
            let lc = Character(c.lowercased())
            if let prev = out.last, Character(prev.lowercased()) == lc, !vowels.contains(lc) {
                if c.isUppercase { out[out.count - 1] = c }
                continue
            }
            out.append(c)
        }
        w = out

        if !prefix {
            // Whole words only. Nasalisation and a trailing bare h carry no
            // information: chhu/chhun/chun, ha/haan/han.
            while w.count > 2, w[w.count - 1] == "n" || w[w.count - 1] == "m", vowels.contains(w[w.count - 2]) {
                w.removeLast()
            }
            while w.count > 2, w[w.count - 1] == "h" {
                w.removeLast()
            }
        }

        return w
    }
}
